### PR TITLE
Show estimates for SELECT query

### PR DESCRIPTION
**Pull request by fastio**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Show estimates for SELECT query. #23941

Detailed description / Documentation draft:

Use case:

```
EXPLAIN ESTIMATES
SELECT
    toYear(LO_ORDERDATE) AS year,
    S_CITY,
    P_BRAND,
    sum(LO_REVENUE - LO_SUPPLYCOST) AS profit
FROM lineorder_flat
WHERE (S_NATION = 'UNITED STATES') AND ((year = 1997) OR (year = 1998)) AND (P_CATEGORY = 'MFGR#14')
GROUP BY
    year,
    S_CITY,
    P_BRAND
ORDER BY
    year ASC,
    S_CITY ASC,
    P_BRAND ASC

Query id: 2ac0e792-6a4c-4c09-95f0-b4d4d9a630c5

┌─database─┬─table──────────┬─parts─┬─────rows─┬─marks─┐
│ default  │ lineorder_flat │    14 │ 14430068 │  1780 │
└──────────┴────────────────┴───────┴──────────┴───────┘
```

### CODE REVIEW

**Comment by UnamedRus** [comment]
I also would like to see count of ranges. (Probably even marks_id?)
And how much each index/partition by/skip index contributed to amount of read rows.


And if you have some nested query like

```
SELECT * FROM table WHERE id IN (SELECT id FROM table_2 WHERE x = 1)
```

In order to get estimates for each level of sub query, you actually need to execute at least some stages of this query.
Probably you can have some setting like this:
```
EXPLAIN ESTIMATES analyze=1 SELECT ...
```

File: src/Interpreters/InterpreterExplainQuery.cpp [hunk 1]
```diff
@@ -283,7 +302,22 @@ BlockInputStreamPtr InterpreterExplainQuery::executeImpl()
         else
             plan.explainPlan(buf, settings.query_plan_options);
     }
-    else if (ast.getKind() == ASTExplainQuery::QueryPipeline)
+    else if (kind == ASTExplainQuery::QueryEstimates)
+    {
+        if (!dynamic_cast<const ASTSelectWithUnionQuery *>(ast.getExplainedQuery().get()))
+            throw Exception("Only SELECT is supported for EXPLAIN query", ErrorCodes::INCORRECT_QUERY);
```
Comment (akuzm) [inline src/Interpreters/InterpreterExplainQuery.cpp:308]:
```suggestion
            throw Exception("Only SELECT is supported for EXPLAIN ESTIMATES query", ErrorCodes::INCORRECT_QUERY);
```

File: src/Processors/QueryPlan/QueryPlan.cpp [hunk 2]
```diff
@@ -429,6 +430,65 @@ void QueryPlan::explainPipeline(WriteBuffer & buffer, const ExplainPipelineOptio
...
+    using CountersPtr = std::shared_ptr<QueryPlan::EstimateCounters>;
+    std::unordered_map<std::string, CountersPtr> counters;
+    std::stack<Frame> stack;
+    stack.push(Frame{.node = root});
+
+    while (!stack.empty())
+    {
+        auto & frame = stack.top();
+
+        if (!frame.is_description_printed)
```
Comment (akuzm) [inline src/Processors/QueryPlan/QueryPlan.cpp:453]: This is somewhat confusing. A 'frame' is just a plan node, right? And this is a depth-first traversal. It should visit all the nodes once, why do you need a special flag to ensure this? A recursive function is probably good enough, these plans are not so deep.

File: src/Processors/QueryPlan/QueryPlan.h [hunk 3]
```diff
@@ -101,7 +103,17 @@ class QueryPlan
     };
 
     using Nodes = std::list<Node>;
-
+    struct EstimateCounters
```
Comment (akuzm) [inline src/Processors/QueryPlan/QueryPlan.h:106]: This is only used by QueryPlan.cpp, so it doesn't need global visibility.

File: src/Processors/QueryPlan/QueryPlan.cpp (hunk 2 above)
Comment (fastio) [reply src/Processors/QueryPlan/QueryPlan.cpp:453]: Ok, we just traverse the plan tree. I will fix it.

**Comment by akuzm** [comment]
We also need a test for this. In case you can't get a stable result for the query, try `explain ... format Null` -- it won't print anything, but at least the query will be run.

**Comment by fastio** [comment]
> We also need a test for this. In case you can't get a stable result for the query, try `explain ... format Null` -- it won't print anything, but at least the query will be run.

@akuzm Thanks for your  reply. I'll fix the problem and add tests soon.

**Comment by fastio** [comment]
Hi @akuzm I close this pr, and submit a new one with tests.

**Comment by fastio** [comment]
Hi @akuzm I close this pr, and submit a new one.

**Pull request closed without merging**